### PR TITLE
[WIP] This PR adds minitest tests.

### PR DESCRIPTION
**Pull request by chaitanyanettem**
I still have a few things to add. I thought I'd show this in case I'm on the wrong track (I'm new to ruby).
- [x] Check if compulsory fields are present in front matter.
- [x] Check if country code is valid.
- [x] Check if there are any html tags in the markdown.
- [x] Integrate with CI

### CODE REVIEW

**Comment by chaitanyanettem** [comment]
This pull request is for #35

**Comment by simison** [comment]
WIP = work in progress aka "Don't merge yet". :-)

**Comment by chaitanyanettem** [comment]
:+1:

**Comment by chaitanyanettem** [comment]
I have finished writing the tests.

The code is located in the `_script/`. I have a weird error with rake wherein the tests run fine but after they are finished rake aborts with this message:

```
rake aborted!
Command failed with status (1): [ruby -I"lib:test"  "/home/arrayjumper/.rvm/rubies/ruby-2.2.1/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/test*.rb" ]

Tasks: TOP => default => test
(See full trace by running task with --trace)
```

~~It'd be great if someone could do these two things. (Your present working directory must be the `_script` folder):~~
- ~~Try running the tests with rake on your machine and see if you also have the same issue. You'll have to issue the command `rake`~~
- ~~Also try running the tests directly like so `ruby -I test:lib test.rb`~~

File: _script/country_codes.txt [hunk 1]
```diff
@@ -0,0 +1,249 @@
+AF
```
Comment (captn3m0) [inline _script/country_codes.txt]: We have this already in the `_data` directory. Use that instead.

File: _script/lib/verify_couches.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+class Couches 
+    def initialize(location = "../_couches")
+        @couch_dir = location
```
Comment (captn3m0) [inline _script/lib/verify_couches.rb]: You can use [glob](http://ruby-doc.org/core-2.0.0/Dir.html#method-c-glob) to get a list of just markdown files.

File: _script/lib/verify_couches.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    def print_couches()
+        @couchnames.each do |couch|
+            puts couch
+        end
+    end
+
+    def separate_YAML_content()
+        @couchnames.each do |couch|
+            couch = File.open("./#{@couch_dir}/#{couch}", "rb").read
+            front_matter = couch[/\A---(.|\n)*?---/]
```
Comment (captn3m0) [inline _script/lib/verify_couches.rb]: This is bad. Use a gem to read front-matter.

File: _script/test.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+require 'minitest/autorun'
```
Comment (captn3m0) [inline _script/test.rb]: test_couches.rb might be a better name

File: _script/test.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                assert(@country_codes.include?(code), "Invalid country code: '#{code_line}' found in #{@couches.couchnames[i]}.\nMake sure it is a valid 2 character country code in CAPS.")
+            end
+            i+=1
+        end
+    end
+
+    def test_content
+        i = 0
+        @couches.all_content.each do |content|
+            matches = content.scan(/<[Hh]\d>/) #Checking for HTML tags that look like <h1>
```
Comment (captn3m0) [inline _script/test.rb]: This is bad as well. Use nokogiri to parse html. See [this answer](http://stackoverflow.com/a/1732454/368328) on why.

**Comment by chaitanyanettem** [comment]
Thanks a ton for that feedback. I wasn't sure whether I should use gems or write my own methods for parsing. So I decided to write my own as an exercise. I'll fix these asap.

File: _script/test_couches.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+class TestCouches < Minitest::Test
+    def setup
+        @couches = Couches.new
+        @country_codes = YAML.load_file("./_data/countries.yml")
+    end
+
+    def test_compulsory_fields
+        i = 0
+        @couches.all_front_matter.each do |front_matter|
+            assert((front_matter.class == Hash), "Incorrect format in front matter in #{@couches.couchnames[i]}")
```
Comment (captn3m0) [inline _script/test_couches.rb]: http://ruby-doc.org/stdlib-2.0.0/libdoc/minitest/rdoc/MiniTest/Assertions.html#method-i-assert_kind_of

File: _script/test_couches.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            assert(front_matter.has_key?("city"), "City field not found in #{@couches.couchnames[i]}")
+            assert(front_matter.has_key?("country"), "Country field not found in #{@couches.couchnames[i]}")
+            assert(front_matter.has_key?("email"), "Email field not found in #{@couches.couchnames[i]}")
+            i+=1
+        end 
+    end
+
+    def test_country_code
+        i = 0
```
Comment (captn3m0) [inline _script/test_couches.rb]: Why is this counter here?

File: _script/test_couches.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            end
+            i+=1
+        end
+    end
+
+    def test_content
+        i = 0
+        @couches.all_content.each do |content|
+            html_fragment = Nokogiri::HTML.fragment(content)
+            matches = html_fragment.search("h1, h2, h3, h4, h5, h6, p, li, ul") #Checking for HTML tags
```
Comment (captn3m0) [inline _script/test_couches.rb]: We only disallow h1 tags. Rest all should be good.

File: _script/test_couches.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            assert(matches.size==0, "Found HTML tags in #{@couches.couchnames[i]}")
```
Comment (captn3m0) [inline _script/test_couches.rb]: We should instead give a better error message (exact offending tag + tag's inner content would be good)

File: _script/test_couches.rb (hunk 8 above)
Comment (simison) [inline _script/test_couches.rb]: So we'll need a list of tags which are okay and the rest should be wiped out, I recon? Same for the html attributes.

Comment (captn3m0) [inline _script/test_couches.rb]: I'd want to run a sanitizer on top of this. https://github.com/flavorjones/loofah is a pretty good library that I've used in the past that handles this.

Things I don't want to let people do:
- Add their own styles
- Use h1 tags
- XSS things

Rest all should be fine for now, and we can look into things as people try them out. (`nav` for eg).

Comment (simison) [inline _script/test_couches.rb]: > Rest all should be fine for now, and we can look into things as people try them out. (nav for eg).

Wouldn't it be easier just to define a whitelist of tags and attributes?

Here's one set I've been using: https://github.com/Trustroots/trustroots/blob/master/modules/core/server/controllers/text-processor.server.controller.js#L9-L40

Comment (captn3m0) [inline _script/test_couches.rb]: That list looks good enough, actually. Covers most of the things that markdown would do.

File: _script/test_couches.rb (hunk 7 above)
Comment (chaitanyanettem) [inline _script/test_couches.rb]: The counter is there to keep track of the _current_ couch being tested. So the test can print the name of the errant couch.

File: _script/test_couches.rb (hunk 6 above)
Comment (chaitanyanettem) [inline _script/test_couches.rb]: Will change.

File: _script/test_couches.rb (hunk 9 above)
Comment (chaitanyanettem) [inline _script/test_couches.rb]: Will fix :+1:

**Comment by captn3m0** [comment]
@chaitanyanettem Any updates on this?

**Comment by chaitanyanettem** [comment]
@captn3m0 I have been travelling. Just got back today. Will have some progress to show tonight.

**Comment by chaitanyanettem** [comment]
@captn3m0 Hey. I have been looking to find a ruby gem that would take care of splitting the couches into front matter and content but I haven't been able to. This is the code I wrote for doing the splitting.

``` ruby
    def separate_YAML_content()
        @couchnames.each do |couch|
            couch = File.open("#{couch}", "r").read
            front_matter = couch[/\A---(.|\n)*?---/]
            content = couch[front_matter.length..couch.length]
            front_matter_hash = YAML.load(front_matter)
            @all_front_matter.push(front_matter_hash)
            @all_content.push(content)
        end
    end
```

I think it should be fine. I'm not doing any parsing per se. Just finding out where the second `---` is and splitting the file at that point. Suggestions?

**Comment by captn3m0** [comment]
@chaitanyanettem I found https://github.com/waiting-for-dev/front_matter_parser

**Comment by chaitanyanettem** [comment]
@captn3m0 That library is basically doing the same thing: https://github.com/waiting-for-dev/front_matter_parser/blob/master/lib/front_matter_parser.rb

I'll use it if you want. But I don't see why that would be better. It is also using a regex to split the front matter and content.

**Comment by captn3m0** [comment]
I'd still argue that

``` ruby
parsed = FrontMatterParser.parse_file(couch)
front_matter = parsed.front_matter
content = parsed.content
```

Is better than 

```
couch = File.open("#{couch}", "r").read
front_matter = couch[/\A---(.|\n)*?---/]
content = couch[front_matter.length..couch.length]
front_matter_hash = YAML.load(front_matter)
```

Also, string interpolation is unnecessary in the first line here. Instead of three variables (couchnames, front_matter array, content array), can we instead make a single array of hash?

``````
[{
"file"=>"whatever.md",
"content"=>"",
"frontmatter"=>{}
},...]```
``````

**Comment by chaitanyanettem** [comment]
Ok. This pull request has dragged on for too long. This ends today.

Success

I'll add the front_matter_parser lib and look at changing to a single array of hash.

**Comment by captn3m0** [comment]
:+1:

**Comment by chaitanyanettem** [comment]
I had some internet issues. Just pushed a commit from the night. Hope it looks okay.

**Comment by captn3m0** [comment]
Rebase this against the current master? Seems like it has conflicts.

**Comment by chaitanyanettem** [comment]
So I squashed my commits and rebased against current master. 

Looks like it can't find `front_matter_parser`. Any suggestions?

**Comment by captn3m0** [comment]
Looking into it, will hopefully merge this by tonight.

**Comment by chaitanyanettem** [comment]
:+1:

**Comment by chaitanyanettem** [comment]
@captn3m0 any updates?

**Comment by captn3m0** [comment]
I think the issue was the gem and ruby versions conflicting. I'd gotten it working somehow. Let me lookup my local branch

**Comment by captn3m0** [comment]
Closing this in favor of the local branch PR: #93

**Pull request closed without merging**